ceph_iscsi_config/lun.py

```diff
@@ -1016,7 +1013,18 @@ def define_luns(logger, config, gateway):
                                     raise CephiSCSIError("Error defining rbd "
                                                          "image {}".format(disk_key))
 
-                                lun.allocate()
+                                disk_config = config.config['disks'][disk_key]
+                                is_allocating_host = local_gw == disk_config['allocating_host']
+                                is_lun_mapped = False
+                                for _, target_config in config.config['targets'].items():
+                                    if local_gw in target_config['portals'] \
+                                            and disk_key in target_config['disks']:
+                                        is_lun_mapped = True
+                                        break
+
+                                if is_allocating_host or is_lun_mapped:
```

Comment:
Does this prevent the creation of an associated storage object if the LUN isn't mapped to a target or the current node isn't the original node to create the disk? If so, I like the idea of deferring the creation of the backing storage object. However, I wonder if we could also eliminate the need for the allocating host and always allow the local ```rbd-target-api``` to delete the disk upon request (since the storage object should be removed upon the removal of the last mapping).

Reply:
> Does this prevent the creation of an associated storage object if the LUN isn't mapped to a target or the current node isn't the original node to create the disk?

Yes, this is exactly what this is doing

> However, I wonder if we could also eliminate the need for the allocating host...

I think that we always have to allocate one backing storage object for each disk declared in the configuration file to prevent the user from deleting the RBD image directly from ceph and end up with an inconsistent `gateway.conf` file that contains disks pointing to RBD images that no longer exist. Am I wrong?

Reply:
Good point. My concern was about the potential awkwardness of telling a ```gwcli``` user that they had to go back to the original host they used to create the image to remove it. If we just don't define the storage object and someone deletes it, the ```LUN.define_luns``` method could be tweaked to ignore ```rbd.ImageNotFound``` in the case where the disk doesn't have any mappings -- and the UI could display the disk attributes as a red ```NOT FOUND``` in the UI tree.

Reply:
@dillaman I've pushed a commit that will only create LIO storage objects when needed, so the user will be able to delete unmapped disks from any host.